diff --git a/zwavelib.cpp b/zwavelib.cpp
index 255d4fb..0a67656 100644
--- a/zwavelib.cpp
+++ b/zwavelib.cpp
@@ -225,6 +225,8 @@ const char *cclassStr (uint8 cc)
     return "DOOR LOCK";
   case 0x63:
     return "USER CODE";
+  case 0x66:
+    return "BARRIER OPERATOR";
   case 0x70:
     return "CONFIGURATION";
   case 0x71:
@@ -393,6 +395,8 @@ uint8 cclassNum (char const *str)
     return 0x62;
   else if (strcmp(str, "USER CODE") == 0)
     return 0x63;
+  else if (strcmp(str, "BARRIER OPERATOR") == 0)
+    return 0x66;
   else if (strcmp(str, "CONFIGURATION") == 0)
     return 0x70;
   else if (strcmp(str, "ALARM") == 0)
